### PR TITLE
[CNXC-281] Optimize feed fetching

### PR DESCRIPTION
**Pull request by janakitti**
## Problem Statement
The results table takes a very long to time to load a single page (upwards of 10 seconds). This very inconvenient for users, especially when the entire table needs to refresh whenever new results are available.

## Solution
Instead of fetching all of the results data to populate a page, the UI should only fetch the data necessary to display the top-level rows. The UI should fetch the nested results when the user chooses to expand a given row.

## Implementation
This implementation separates the fetching of data for the parent rows from the nested results. When an analysis is run, a Note resource is created along with the Feed containing a stringified JSON of the parent row data. When Feeds are fetched to display the table, the Notes are read to populate the parent rows. When the user expands a row, the results data is fetches and loaded into the SeriesTable.

### CODE REVIEW

File: src/app.config.tsx [hunk 1]
```diff
@@ -21,5 +21,10 @@ export const PluginModels: Plugins = {
 // The base filepath for PACS files in the Swift storage
 export const BASE_PACS_FILE_PATH = 'SERVICES/PACS/covidnet/';
 
+// Note title for when creating pl-dircopy instance
+export const FEED_NOTE_TITLE = 'COVIDNET_ANALYSIS_NOTE';
```
Comment (boris-nguyen) [inline src/app.config.tsx:25]:
```suggestion
export const FEED_NOTE_TITLE = "COVIDNET_ANALYSIS_NOTE";
```

File: src/components/pastAnalysis/seriesTable.tsx [hunk 2]
```diff
@@ -99,19 +110,31 @@ const SeriesTable: React.FC<SeriesTableProps> = ({ studyInstance, classification
...
+      (
+          <Table aria-label="Simple Table" cells={columns} rows={rows}>
+            <TableHeader />
+            <TableBody className="series-table-row" />
+          </Table>
+      ):
+      (
+        <div className="results-spinner-container">
+          <Spinner size="lg" />
+          </div>
```
Comment (boris-nguyen) [inline src/components/pastAnalysis/seriesTable.tsx:134]:
```suggestion
        </div>
```

**Review by colinywu — APPROVED** [review]
Great work! Lot of high quality work and massive changes that really improve the project and boost performance

File: src/services/CreateAnalysisService.tsx [hunk 3]
```diff
@@ -70,16 +71,17 @@ class CreateAnalysisService {
...
+      return await ChrisIntegration.processOneImg(img, timestamp, XrayModel, CTModel)
     }));
 
     const results = await processedImages.flatMap((img: PromiseSettledResult<BackendPollResult>, index: number) => {
       if (img?.status === "fulfilled" && img?.value?.error) {
         return {
           image: dcmImages[index],
           processedResults: img.value
-        };
+        }
```
Comment (boris-nguyen) [inline src/services/CreateAnalysisService.tsx:84]:
```suggestion
        };
```

File: src/services/chris_integration.tsx [hunk 4]
```diff
@@ -240,19 +282,33 @@ class ChrisIntegration {
...
+   * @returns {Promise<boolean>} True of all Feed jobs are completed
    */
-  static async checkIfPluginTerminated(id: number): Promise<boolean> {
+  static async checkIfFeedJobsCompleted(id: number): Promise<boolean> {
     const client: Client = ChrisAPIClient.getClient();
-    const plugin = await client.getPluginInstances({ id });
-    const status = plugin?.getItems()?.[0]?.data?.status;
-    return status === PluginPollStatus.SUCCESS || status === PluginPollStatus.ERROR || status === PluginPollStatus.CANCELLED;
+    const feed: Feed = await client.getFeed(id);
+    const feedData = feed?.data
```
Comment (boris-nguyen) [inline src/services/chris_integration.tsx:292]:
```suggestion
    const feedData = feed?.data;
```

File: src/services/chris_integration.tsx [hunk 5]
```diff
@@ -262,160 +318,138 @@ class ChrisIntegration {
...
+   * @param {number[]} feedIds List of Feed IDs
+   * @return {Promise<TAnalysisResults>} Results from analysis
+   */
+  static async getResultsAndClassesFromFeedIds(feedIds: number[]): Promise<TAnalysisResults> {
+    const series: ISeries[] = await Promise.all(feedIds.map(async (id: number): Promise<ISeries> => {
+      const covidnetPlugin = await this.getCovidnetPluginInstanceFromFeedId(id);
+      return await this.getCovidnetResults(covidnetPlugin);
+    }));
+    
+    return {series, classifications: Array.from(series[0].classifications.keys())}
```
Comment (boris-nguyen) [inline src/services/chris_integration.tsx:430]:
```suggestion
    return {series, classifications: Array.from(series?.[0]?.classifications.keys())}
```

File: src/services/chris_integration.tsx [hunk 6]
```diff
@@ -262,160 +318,138 @@ class ChrisIntegration {
...
+          note: noteContent
+        }];
+      }))).flat();
+
+      feedNoteArray.push(...newFeedNoteArray);
 
-      // Update fetchLimit to be remaining number of StudyInstanceWithSeries desired
-      fetchLimit = limit + 1 - pastAnalysis.length;
+      // Group Feeds into an object by [timestamp, StudyInstanceUID]. Each group represents a row on the table.
+      studyGroups = groupBy(feedNoteArray, (feedNote: any) => [feedNote.note.timestamp, feedNote.note.img.StudyInstanceUID]);
```
Comment (boris-nguyen) [inline src/services/chris_integration.tsx:371]:
```suggestion
      studyGroups = groupBy(feedNoteArray, (feedNote: TFeedNote) => [feedNote.note.timestamp, feedNote.note.img.StudyInstanceUID]);
```

File: src/services/chris_integration.tsx [hunk 7]
```diff
@@ -262,160 +318,138 @@ class ChrisIntegration {
...
+        pluginStatuses
+      }
+    });
+
+    // Discard the extra analysis, if there is one
+    const pastAnalysesSliced = pastAnalyses.slice(0, limit);
+
+    // Count the number of Feeds that are actually being used for this page and add to the initial offset
+    const lastOffset = offset + pastAnalysesSliced.reduce((acc: number, cur: TStudyInstance) => {
+      return acc + cur.feedIds.length
```
Comment (boris-nguyen) [inline src/services/chris_integration.tsx:413]:
```suggestion
      return acc + cur.feedIds.length;
```

**Pull request merged**